cmd/otel-allocator/prehook/relabel.go

```diff
@@ -0,0 +1,108 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prehook
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/target"
+)
+
+type RelabelConfigTargetFilter struct {
+	m          sync.RWMutex
+	log        logr.Logger
+	relabelCfg map[string][]*relabel.Config
+}
+
+func NewRelabelConfigTargetFilter(log logr.Logger) Hook {
+	return &RelabelConfigTargetFilter{
+		log:        log,
+		relabelCfg: make(map[string][]*relabel.Config),
+	}
+}
+
+// helper function converts from model.LabelSet to []labels.Label.
+func convertLabelToPromLabelSet(lbls model.LabelSet) []labels.Label {
+	newLabels := make([]labels.Label, len(lbls))
+	index := 0
+	for k, v := range lbls {
+		newLabels[index].Name = string(k)
+		newLabels[index].Value = string(v)
+		index++
+	}
+	return newLabels
+}
+
+func (tf *RelabelConfigTargetFilter) Apply(targets map[string]*target.Item) map[string]*target.Item {
+	numTargets := len(targets)
+	numRemainingTargets := numTargets
+
+	// need to wait until relabelCfg is set
+	if len(tf.relabelCfg) == 0 {
+		return targets
+	}
+	// Note: jobNameKey != tItem.JobName (jobNameKey is hashed)
+	for jobNameKey, tItem := range targets {
+		keepTarget := true
+		lset := convertLabelToPromLabelSet(tItem.Label)
+		for _, cfg := range tf.relabelCfg[tItem.JobName] {
+			if new_lset := relabel.Process(lset, cfg); new_lset == nil {
+				keepTarget = false
+				break // inner loop
+			} else {
+				lset = new_lset
+			}
+		}
+
+		if !keepTarget {
+			delete(targets, jobNameKey)
+			numRemainingTargets--
+		}
+	}
+
+	tf.log.V(2).Info("Filtering complete", "seen", numTargets, "kept", numRemainingTargets)
+	return targets
+}
+
+func (tf *RelabelConfigTargetFilter) SetConfig(cfgs map[string][]*relabel.Config) {
+	tf.m.Lock()
+	defer tf.m.Unlock()
+
+	for key, val := range cfgs {
+		tf.relabelCfg[key] = val
```

Comment:
This reads more like it is merging configs than setting a new config, as the old config entries that are not present in the new config will remain unchanged.

Reply:
Now I am creating a new cfg and assigning its value to tf.relabelCfg, so stale keys will no longer be present. Thanks for pointing this out!